runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java

```diff
@@ -0,0 +1,197 @@
+package org.corfudb.protocols.wireprotocol;
+
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.timeout.ReadTimeoutException;
+import io.netty.handler.timeout.ReadTimeoutHandler;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * The ClientHandshakeHandler initiates the handshake upon socket connection.
+ *
+ * - Once the client connects to the server, it sends a handshake message that contains:
+ *         its own id and the (asserted) server's node id.
+ * - The server validates and replies with its node id and current version of Corfu.
+ * - If validation is correct on both sides, message exchange is initiated between client-server,
+ * otherwise, the handshake times out, and either server or client close the connection.
+ *
+ * Created by amartinezman on 12/8/17.
+ */
+@Slf4j
+public class ClientHandshakeHandler extends ChannelDuplexHandler {
+
+    private final UUID clientId;
+    private final UUID nodeId;
+    private final AtomicBoolean handshakeComplete;
```

Comment:
As before, you might enter a state in which handshake hasn't failed (the default value) but not because it actually completed but because nothing has happened yet. Or it might have failed not because it completed but because it timed out.

Reply:
So the variable you use is an object. It has a third "construed" state: uninitialized which can be used for that, But as long as you are comfortable, I am ok. My suggestion would be to couple these variables together in one object so that they mutate together and there is less chance of mistakes for later as well.